<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>md4mg: metadata for metagenomics • md4mg</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="md4mg: metadata for metagenomics">
<meta property="og:description" content="md4mg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">md4mg</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/md4mg.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/vjcitn/md4mg">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="md4mg_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="md4mg_files/htmlwidgets-1.5.1/htmlwidgets.js"></script><script src="md4mg_files/jquery-1.12.4/jquery.min.js"></script><link href="md4mg_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="md4mg_files/datatables-binding-0.14/datatables.js"></script><link href="md4mg_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="md4mg_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="md4mg_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="md4mg_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="md4mg_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>md4mg: metadata for metagenomics</h1>
                        <h4 class="author">Vince Carey<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
                        <h4 class="author">Sean Davis<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/md4mg/blob/master/vignettes/md4mg.Rmd"><code>vignettes/md4mg.Rmd</code></a></small>
      <div class="hidden name"><code>md4mg.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This vignette reviews basic aspects of interacting with omicidx to obtain metadata about metagenomics studies available in NCBI SRA.</p>
<div id="a-quick-view-of-a-query-about-a-study" class="section level2">
<h2 class="hasAnchor">
<a href="#a-quick-view-of-a-query-about-a-study" class="anchor"></a>A quick view of a query about a study</h2>
<p>To get things rolling, let’s assume we know the identifier for an SRA <em>study</em> of interest: SRP114847. How we discover identifiers for studies or families of studies will be discussed later.</p>
<p>We can obtain a rectangular table of information about SRP114847 by querying a specific <em>endpoint</em> of the omicidx API. We’ll use the “runs for study” endpoint and then discuss why and how afterwards.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tibble</span>)
<span class="no">metadata</span><span class="kw">=</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span>(<span class="kw pkg">jsonlite</span><span class="kw ns">::</span><span class="fu"><a href="https://jeroen.cran.dev/jsonlite/reference/fromJSON.html">fromJSON</a></span>(<span class="kw">txt</span><span class="kw">=</span>
   <span class="fu"><a href="https://rdrr.io/r/base/connections.html">url</a></span>(<span class="st">'https://api.omicidx.cancerdatasci.org/sra/studies/SRP114847/runs?size=500'</span>),
   <span class="kw">flatten</span><span class="kw">=</span><span class="fl">TRUE</span>)$<span class="no">hits</span>)
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">metadata</span>)
<span class="co">#&gt; [1] 219  70</span></pre></body></html></div>
<p>So there are 219 rows returned. We can visit <a href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP114847">the study level page at SRA</a> to find:</p>
<p><img src="srapic1.png"></p>
<p>The mention of 219 runs suggests that the API did its job correctly.</p>
</div>
<div id="the-fields-returned" class="section level2">
<h2 class="hasAnchor">
<a href="#the-fields-returned" class="anchor"></a>The fields returned</h2>
<p>On the basis of this query, the <code>metadata</code> table could have no more than 500 rows. What are the fields returned by our query?</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">metadata</span>)
<span class="co">#&gt;  [1] "cluster_name"                 "static_data_available"       </span>
<span class="co">#&gt;  [3] "total_spots"                  "avg_length"                  </span>
<span class="co">#&gt;  [5] "sample_accession"             "experiment_accession"        </span>
<span class="co">#&gt;  [7] "received"                     "accession"                   </span>
<span class="co">#&gt;  [9] "published"                    "study_accession"             </span>
<span class="co">#&gt; [11] "size"                         "insdc"                       </span>
<span class="co">#&gt; [13] "load_done"                    "is_public"                   </span>
<span class="co">#&gt; [15] "alias"                        "qualities"                   </span>
<span class="co">#&gt; [17] "attributes"                   "total_bases"                 </span>
<span class="co">#&gt; [19] "lastupdate"                   "status"                      </span>
<span class="co">#&gt; [21] "study.pubmed_ids"             "study.identifiers"           </span>
<span class="co">#&gt; [23] "study.received"               "study.abstract"              </span>
<span class="co">#&gt; [25] "study.accession"              "study.published"             </span>
<span class="co">#&gt; [27] "study.title"                  "study.center_name"           </span>
<span class="co">#&gt; [29] "study.study_type"             "study.insdc"                 </span>
<span class="co">#&gt; [31] "study.alias"                  "study.attributes"            </span>
<span class="co">#&gt; [33] "study.lastupdate"             "study.BioProject"            </span>
<span class="co">#&gt; [35] "study.status"                 "sample.organism"             </span>
<span class="co">#&gt; [37] "sample.identifiers"           "sample.xrefs"                </span>
<span class="co">#&gt; [39] "sample.description"           "sample.taxon_id"             </span>
<span class="co">#&gt; [41] "sample.received"              "sample.accession"            </span>
<span class="co">#&gt; [43] "sample.published"             "sample.BioSample"            </span>
<span class="co">#&gt; [45] "sample.insdc"                 "sample.alias"                </span>
<span class="co">#&gt; [47] "sample.attributes"            "sample.lastupdate"           </span>
<span class="co">#&gt; [49] "sample.status"                "experiment.library_layout"   </span>
<span class="co">#&gt; [51] "experiment.identifiers"       "experiment.sample_accession" </span>
<span class="co">#&gt; [53] "experiment.xrefs"             "experiment.received"         </span>
<span class="co">#&gt; [55] "experiment.accession"         "experiment.published"        </span>
<span class="co">#&gt; [57] "experiment.title"             "experiment.library_strategy" </span>
<span class="co">#&gt; [59] "experiment.platform"          "experiment.study_accession"  </span>
<span class="co">#&gt; [61] "experiment.insdc"             "experiment.instrument_model" </span>
<span class="co">#&gt; [63] "experiment.design"            "experiment.alias"            </span>
<span class="co">#&gt; [65] "experiment.library_name"      "experiment.attributes"       </span>
<span class="co">#&gt; [67] "experiment.library_selection" "experiment.lastupdate"       </span>
<span class="co">#&gt; [69] "experiment.library_source"    "experiment.status"</span></pre></body></html></div>
<p>There is some structure in the field names. There are four families of fields, base-level, study-level, experiment-level, and sample-level. The study/experiment/sample hierarchy reflects the fundamental <em>data model</em> of NCBI SRA.</p>
<p>It is somewhat curious that the field names do not include the token <code>run</code>. It seems reasonable to say that ‘run’ is a synonym for ‘experiment’, but the relationship is not always one-to-one. In this case, we see that each row returned has a unique experiment accession number:</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">metadata</span>$<span class="no">experiment.accession</span>))
<span class="co">#&gt; [1] 219</span></pre></body></html></div>
</div>
<div id="the-sample-attributes" class="section level2">
<h2 class="hasAnchor">
<a href="#the-sample-attributes" class="anchor"></a>The sample attributes</h2>
<p>A challenging aspect of SRA metadata is the diversity of sample attributes recorded, which varies from study to study. The <code>metadata</code> <code>sample.attributes</code> field is nested within the metadata table, and consists of a list of tag-value pairs. As examples:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">metadata</span>$<span class="no">sample.attributes</span>[<span class="fl">1</span>:<span class="fl">2</span>]
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt;                tag                       value</span>
<span class="co">#&gt; 1  collection_date                 20-Jul-2016</span>
<span class="co">#&gt; 2        env_biome                   Human-gut</span>
<span class="co">#&gt; 3      env_feature    Human associated habitat</span>
<span class="co">#&gt; 4     env_material                      Faeces</span>
<span class="co">#&gt; 5     geo_loc_name               India: Bhopal</span>
<span class="co">#&gt; 6             host                Homo sapiens</span>
<span class="co">#&gt; 7          lat_lon             23.26 N 77.41 E</span>
<span class="co">#&gt; 8           Number                    Sample12</span>
<span class="co">#&gt; 9         host_age                          29</span>
<span class="co">#&gt; 10       host_diet                         Veg</span>
<span class="co">#&gt; 11  BioSampleModel Metagenome or environmental</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt;                tag                       value</span>
<span class="co">#&gt; 1  collection_date                 20-Jul-2016</span>
<span class="co">#&gt; 2        env_biome                   Human-gut</span>
<span class="co">#&gt; 3      env_feature    Human associated habitat</span>
<span class="co">#&gt; 4     env_material                      Faeces</span>
<span class="co">#&gt; 5     geo_loc_name               India: Bhopal</span>
<span class="co">#&gt; 6             host                Homo sapiens</span>
<span class="co">#&gt; 7          lat_lon             23.26 N 77.41 E</span>
<span class="co">#&gt; 8           Number                    Sample11</span>
<span class="co">#&gt; 9         host_age                          22</span>
<span class="co">#&gt; 10       host_diet                         Veg</span>
<span class="co">#&gt; 11  BioSampleModel Metagenome or environmental</span></pre></body></html></div>
<p>This is somewhat cumbersome but it allows for diversity in metadata capture from sample to sample.</p>
<p>To transform the <code>sample.attributes</code> to a table with one row per run, tidyr pivoting is very efficient:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">sampatts</span> <span class="kw">=</span> <span class="no">metadata</span> <span class="kw">%&gt;%</span> <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">accession</span>, <span class="no">sample.attributes</span>) <span class="kw">%&gt;%</span>
    <span class="kw pkg">tidyr</span><span class="kw ns">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span>(<span class="no">sample.attributes</span>) <span class="kw">%&gt;%</span>
    <span class="kw pkg">tidyr</span><span class="kw ns">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span>(<span class="kw">names_from</span><span class="kw">=</span><span class="st">'tag'</span>, <span class="kw">values_from</span><span class="kw">=</span><span class="st">'value'</span>)
<span class="kw pkg">DT</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html">datatable</a></span>(<span class="no">sampatts</span>)</pre></body></html></div>
<p></p>
<div id="htmlwidget-0efff4b49e05d48ae78f" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-0efff4b49e05d48ae78f">{"x":{"filter":"none","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204","205","206","207","208","209","210","211","212","213","214","215","216","217","218","219"],["SRR5898908","SRR5898909","SRR5898910","SRR5898911","SRR5898912","SRR5898913","SRR5898914","SRR5898915","SRR5898916","SRR5898917","SRR5898918","SRR5898919","SRR5898920","SRR5898921","SRR5898922","SRR5898923","SRR5898924","SRR5898925","SRR5898926","SRR5898927","SRR5898928","SRR5898929","SRR5898930","SRR5898931","SRR5898932","SRR5898933","SRR5898934","SRR5898935","SRR5898936","SRR5898937","SRR5898938","SRR5898939","SRR5898940","SRR5898941","SRR5898942","SRR5898943","SRR5898944","SRR5898945","SRR5898946","SRR5898947","SRR5898948","SRR5898949","SRR5898950","SRR5898951","SRR5898952","SRR5898953","SRR5898954","SRR5898955","SRR5898956","SRR5898957","SRR5898958","SRR5898959","SRR5898960","SRR5898961","SRR5898962","SRR5898963","SRR5898964","SRR5898965","SRR5898966","SRR5898967","SRR5898968","SRR5898969","SRR5898970","SRR5898971","SRR5898972","SRR5898973","SRR5898974","SRR5898975","SRR5898976","SRR5898977","SRR5898978","SRR5898979","SRR5898980","SRR5898981","SRR5898982","SRR5898983","SRR5898984","SRR5898985","SRR5898986","SRR5898987","SRR5898988","SRR5898989","SRR5898990","SRR5898991","SRR5898992","SRR5898993","SRR5898994","SRR5898995","SRR5898996","SRR5898997","SRR5898998","SRR5898999","SRR5899000","SRR5899001","SRR5899002","SRR5899003","SRR5899004","SRR5899005","SRR5899006","SRR5899007","SRR5899008","SRR5899009","SRR5899010","SRR5899011","SRR5899012","SRR5899013","SRR5899014","SRR5899015","SRR5899016","SRR5899017","SRR6166631","SRR6166632","SRR6166633","SRR6166634","SRR6166635","SRR6166636","SRR6166637","SRR6166638","SRR6166639","SRR6166640","SRR6166641","SRR6166642","SRR6166643","SRR6166644","SRR6166645","SRR6166646","SRR6166647","SRR6166648","SRR6166649","SRR6166650","SRR6166651","SRR6166652","SRR6166653","SRR6166654","SRR6166655","SRR6166656","SRR6166657","SRR6166658","SRR6166659","SRR6166660","SRR6166661","SRR6166662","SRR6166663","SRR6166664","SRR6166665","SRR6166666","SRR6166667","SRR6166668","SRR6166669","SRR6166670","SRR6166671","SRR6166672","SRR6166673","SRR6166674","SRR6166675","SRR6166676","SRR6166677","SRR6166678","SRR6166679","SRR6166680","SRR6166681","SRR6166682","SRR6166683","SRR6166684","SRR6166685","SRR6166686","SRR6166687","SRR6166688","SRR6166689","SRR6166690","SRR6166691","SRR6166692","SRR6166693","SRR6166694","SRR6166695","SRR6166696","SRR6166697","SRR6166698","SRR6166699","SRR6166700","SRR6166701","SRR6166702","SRR6166703","SRR6166704","SRR6166705","SRR6166706","SRR6166707","SRR6166708","SRR6166709","SRR6166710","SRR6166711","SRR6166712","SRR6166713","SRR6166714","SRR6166715","SRR6166716","SRR6166717","SRR6166718","SRR6166719","SRR6166720","SRR6166721","SRR6166722","SRR6166723","SRR6166724","SRR6166725","SRR6166726","SRR6166727","SRR6166728","SRR6166729","SRR6166730","SRR6166731","SRR6166732","SRR6166733","SRR6166734","SRR6166735","SRR6166736","SRR6166737","SRR6166738","SRR6166739"],["20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016","20-Jul-2016"],["Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut","Human-gut"],["Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat","Human associated habitat"],["Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces","Faeces"],["India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India:Kasaragod","India:Kasaragod","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India: Bhopal","India: Bhopal","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India: Bhopal","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India: Bhopal","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India: Bhopal","India:Kasaragod","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India:Kasaragod","India:Kasaragod","India: Bhopal","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India:Kasaragod","India: Bhopal","India: Bhopal","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India: Bhopal","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India: Bhopal","India: Bhopal","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India: Bhopal","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India:Kasaragod","India:Kasaragod","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India: Bhopal","India:Kasaragod","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India: Bhopal","India:Kasaragod","India: Bhopal","India: Bhopal","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod","India:Kasaragod"],["Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens"],["23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","12.51 N 74.98 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","12.51 N 74.98 E","23.26 N 77.41 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","23.26 N 77.41 E","12.51 N 74.98 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","23.26 N 77.41 E","12.51 N 74.98 E","23.26 N 77.41 E","23.26 N 77.41 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E","12.51 N 74.98 E"],["Sample12","Sample11","Sample10","Sample9","Sample8","Sample7","Sample6","Sample5","Sample4","Sample3","Sample74","Sample75","Sample13","Sample14","Sample15","Sample16","Sample17","Sample18","Sample19","Sample73","Sample61","Sample62","Sample59","Sample60","Sample58","Sample71","Sample56","Sample57","Sample54","Sample55","Sample69","Sample70","Sample67","Sample68","Sample65","Sample66","Sample63","Sample64","Sample1","Sample2","Sample110","Sample109","Sample106","Sample72","Sample108","Sample107","Sample104","Sample103","Sample33","Sample105","Sample97","Sample96","Sample95","Sample21","Sample101","Sample100","Sample99","Sample98","Sample22","Sample102","Sample42","Sample43","Sample40","Sample41","Sample38","Sample39","Sample36","Sample37","Sample34","Sample35","Sample31","Sample32","Sample27","Sample28","Sample29","Sample30","Sample23","Sample24","Sample25","Sample26","Sample84","Sample83","Sample20","Sample76","Sample78","Sample77","Sample80","Sample79","Sample82","Sample81","Sample53","Sample52","Sample49","Sample48","Sample51","Sample50","Sample45","Sample44","Sample47","Sample46","Sample93","Sample94","Sample87","Sample88","Sample85","Sample86","Sample91","Sample92","Sample89","Sample90","Sample31","Sample30","Sample29","Sample28","Sample27","Sample26","Sample25","Sample24","Sample103","Sample32","Sample53","Sample96","Sample97","Sample98","Sample99","Sample100","Sample101","Sample102","Sample45","Sample46","Sample47","Sample48","Sample49","Sample50","Sample51","Sample52","Sample57","Sample56","Sample55","Sample54","Sample61","Sample60","Sample59","Sample58","Sample63","Sample62","Sample21","Sample94","Sample91","Sample90","Sample93","Sample92","Sample87","Sample86","Sample89","Sample88","Sample22","Sample23","Sample84","Sample85","Sample82","Sample83","Sample80","Sample81","Sample78","Sample79","Sample20","Sample77","Sample95","Sample76","Sample75","Sample17","Sample16","Sample15","Sample14","Sample74","Sample73","Sample19","Sample18","Sample8","Sample9","Sample10","Sample11","Sample4","Sample5","Sample6","Sample7","Sample12","Sample13","Sample65","Sample64","Sample67","Sample66","Sample69","Sample68","Sample1","Sample70","Sample3","Sample2","Sample44","Sample43","Sample36","Sample35","Sample38","Sample37","Sample40","Sample39","Sample42","Sample41","Sample110","Sample34","Sample33","Sample72","Sample104","Sample105","Sample108","Sample109","Sample106","Sample107"],["29","22","24","25","23","21","21","27","21","28","24","25","24","23","26","26","25","28","26","30","6","5","12","4","16","9","12","11","7","11","3.5","10","9","5","10","6","7","12","26","27","58","30","50","52","30","27","53","21","29","25","56","50","50","23","60","53","53","58","64","53","28","26","29","30","23","26","30","2.4","22","30","61","58","52","62","50","58","62","57","57","71","26","29","23","23","23","23","20","19","30","30","29","0.8","28","25","0.6","0.6","24","30","25","26","60","50","21","30","28","27","55","55","23","50","61","58","50","62","52","71","57","57","21","58","29","50","56","58","53","53","60","53","24","26","25","25","28","0.6","0.6","0.8","11","12","11","7","6","4","12","16","7","5","23","50","55","50","60","55","21","27","23","30","64","62","26","28","30","29","20","30","23","19","23","23","50","23","25","25","26","26","23","24","30","26","28","23","25","24","22","21","27","21","21","29","24","10","12","9","6","3.5","5","26","10","28","27","30","26","30","30","23","2.4","29","26","28","30","58","22","29","52","53","25","30","30","50","27"],["Veg","Veg","Veg","Veg","Veg","Veg","Veg","Veg","Veg","Veg","Non_veg","Non_veg","Veg","Veg","Veg","Veg","Veg","Veg","Veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Veg","Non_veg","Non_veg","Non_veg","Non_veg","Veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Veg","Veg","Non_veg","Veg","Veg","Non_veg","Veg","Veg","Veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Veg","Veg","Veg","Veg","Veg","Veg","Non_veg","Veg","Veg","Veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Veg","Veg","Veg","Veg","Veg","Veg","Veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Veg","Veg","Veg","Veg","Non_veg","Non_veg","Veg","Veg","Veg","Veg","Veg","Veg","Veg","Veg","Veg","Veg","Veg","Veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Veg","Non_veg","Veg","Veg","Non_veg","Veg","Veg","Veg","Veg","Veg","Non_veg","Non_veg","Non_veg","Veg","Veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg","Non_veg"],["Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental","Metagenome or environmental"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>accession<\/th>\n      <th>collection_date<\/th>\n      <th>env_biome<\/th>\n      <th>env_feature<\/th>\n      <th>env_material<\/th>\n      <th>geo_loc_name<\/th>\n      <th>host<\/th>\n      <th>lat_lon<\/th>\n      <th>Number<\/th>\n      <th>host_age<\/th>\n      <th>host_diet<\/th>\n      <th>BioSampleModel<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script> With this interactive table we can quickly see that there is diversity in geographic location (sort on <code>lat_lon</code>), and diet (sort on <code>host_diet</code>), but not on <code>env_material</code> (sort on it).
<p>So, with very little code, we’ve learned a lot about SRP114847. Now let’s learn more about the API.</p>
</div>
</div>
<div id="reflectance-in-omicidx" class="section level1">
<h1 class="hasAnchor">
<a href="#reflectance-in-omicidx" class="anchor"></a>Reflectance in OmicIDX</h1>
<div id="the-basic-idea-listing-the-available-endpoints" class="section level2">
<h2 class="hasAnchor">
<a href="#the-basic-idea-listing-the-available-endpoints" class="anchor"></a>The basic idea – listing the available endpoints</h2>
<p>The OmicIDX system is developed using OpenAPI 3.0. It is possible to learn about the features of OmicIDX by parsing a JSON document.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">ref</span>  <span class="kw">=</span> <span class="kw pkg">jsonlite</span><span class="kw ns">::</span><span class="fu"><a href="https://jeroen.cran.dev/jsonlite/reference/fromJSON.html">fromJSON</a></span>(<span class="st">"https://api.omicidx.cancerdatasci.org/openapi.json"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">ref</span>)
<span class="co">#&gt; [1] "openapi"    "info"       "paths"      "components"</span></pre></body></html></div>
<p>The <code>paths</code> element is really useful for programming with the API.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">ref</span>$<span class="no">paths</span>)
<span class="co">#&gt;  [1] "/sra/fields/{entity}"                </span>
<span class="co">#&gt;  [2] "/sra/studies/{accession}"            </span>
<span class="co">#&gt;  [3] "/sra/studies"                        </span>
<span class="co">#&gt;  [4] "/sra/studies/{accession}/samples"    </span>
<span class="co">#&gt;  [5] "/sra/studies/{accession}/experiments"</span>
<span class="co">#&gt;  [6] "/sra/studies/{accession}/runs"       </span>
<span class="co">#&gt;  [7] "/sra/samples/{accession}"            </span>
<span class="co">#&gt;  [8] "/sra/samples"                        </span>
<span class="co">#&gt;  [9] "/sra/samples/{accession}/experiments"</span>
<span class="co">#&gt; [10] "/sra/samples/{accession}/runs"       </span>
<span class="co">#&gt; [11] "/sra/experiments/{accession}"        </span>
<span class="co">#&gt; [12] "/sra/experiments"                    </span>
<span class="co">#&gt; [13] "/sra/experiments/{accession}/runs"   </span>
<span class="co">#&gt; [14] "/sra/runs/{accession}"               </span>
<span class="co">#&gt; [15] "/sra/runs"                           </span>
<span class="co">#&gt; [16] "/biosample/fields/{entity}"          </span>
<span class="co">#&gt; [17] "/biosample/samples"                  </span>
<span class="co">#&gt; [18] "/biosample/samples/{accession}"      </span>
<span class="co">#&gt; [19] "/facets/{entity}"</span></pre></body></html></div>
<p>It has its own substructure, which you can explore using an expandable tree – try <code>listviewer::jsonedit(ref)</code>. For now we’ll focus on the endpoint we used.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">ourendpt</span> <span class="kw">=</span> <span class="st">"/sra/studies/{accession}/experiments"</span>
<span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="kw">qualname</span><span class="kw">=</span><span class="no">ref</span>$<span class="no">paths</span><span class="kw">[[</span><span class="no">ourendpt</span>]]$<span class="no">get</span>$<span class="no">parameters</span>$<span class="no">name</span>,
  <span class="no">ref</span>$<span class="no">paths</span><span class="kw">[[</span><span class="no">ourendpt</span>]]$<span class="no">get</span>$<span class="no">parameters</span>$<span class="no">schema</span>))  <span class="co"># wanted datatable but threw warning</span>
<span class="co">#&gt;         qualname          title    type</span>
<span class="co">#&gt; 1      accession      Accession  string</span>
<span class="co">#&gt; 2 include_fields Include_Fields   array</span>
<span class="co">#&gt; 3 exclude_fields Exclude_Fields   array</span>
<span class="co">#&gt; 4           size           Size integer</span>
<span class="co">#&gt; 5         cursor         Cursor  string</span>
<span class="co">#&gt;                                                                  description</span>
<span class="co">#&gt; 1                                                    An accession for lookup</span>
<span class="co">#&gt; 2             Fields to include in results. The default is to all fields (*)</span>
<span class="co">#&gt; 3 Fields to exclude from results. The default is to not exclude any fields. </span>
<span class="co">#&gt; 4                                                                       &lt;NA&gt;</span>
<span class="co">#&gt; 5                                                                       &lt;NA&gt;</span>
<span class="co">#&gt;     type default example exclusiveMaximum</span>
<span class="co">#&gt; 1   &lt;NA&gt;    NULL    NULL               NA</span>
<span class="co">#&gt; 2 string    NULL       *               NA</span>
<span class="co">#&gt; 3 string    NULL    NULL               NA</span>
<span class="co">#&gt; 4   &lt;NA&gt;      10      10             1000</span>
<span class="co">#&gt; 5   &lt;NA&gt;    NULL    NULL               NA</span></pre></body></html></div>
</div>
<div id="trying-out-a-different-endpoint" class="section level2">
<h2 class="hasAnchor">
<a href="#trying-out-a-different-endpoint" class="anchor"></a>Trying out a different endpoint</h2>
<p>We found that <code>refs$paths</code> has lots of information that our software can use to figure out what it can do with the OmicIDX API. Here we’ll add a lucene query for the sra/studies endpoint. We’ll modify the default number of hits to retrieve, which is 10, specifying that we want up to 15 records. We’ll use <code>q=microbiome%20AND%20cancer</code> as a Lucene query.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">mcq</span> <span class="kw">=</span> <span class="kw pkg">jsonlite</span><span class="kw ns">::</span><span class="fu"><a href="https://jeroen.cran.dev/jsonlite/reference/fromJSON.html">fromJSON</a></span>(
 <span class="st">"https://api.omicidx.cancerdatasci.org/sra/studies?q=microbiome%20AND%20cancer&amp;size=15"</span>,
 <span class="kw">flatten</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span>(<span class="no">mcq</span>$<span class="no">hits</span>)
<span class="co">#&gt; # A tibble: 15 x 23</span>
<span class="co">#&gt;    pubmed_ids total_spots identifiers experiment_count received abstract</span>
<span class="co">#&gt;    &lt;list&gt;     &lt;chr&gt;       &lt;list&gt;      &lt;chr&gt;            &lt;chr&gt;    &lt;chr&gt;   </span>
<span class="co">#&gt;  1 &lt;list [0]&gt; 2669228     &lt;df[,2] [2… 18               2020-04… Argan p…</span>
<span class="co">#&gt;  2 &lt;list [0]&gt; 26923       &lt;df[,2] [3… 2                2013-07… Objecti…</span>
<span class="co">#&gt;  3 &lt;list [0]&gt; 2743023     &lt;df[,2] [2… 41               2014-11… This st…</span>
<span class="co">#&gt;  4 &lt;list [0]&gt; 117647491   &lt;df[,2] [2… 96               2014-11… Individ…</span>
<span class="co">#&gt;  5 &lt;list [0]&gt; 6920681858  &lt;df[,2] [2… 1515             2014-11… Several…</span>
<span class="co">#&gt;  6 &lt;list [0]&gt; 495190      &lt;df[,2] [2… 39               2015-05… Backgro…</span>
<span class="co">#&gt;  7 &lt;list [0]&gt; 4262914461  &lt;df[,2] [2… 312              2015-03… Colorec…</span>
<span class="co">#&gt;  8 &lt;list [0]&gt; 27646       &lt;df[,2] [2… 1                2015-02… Heat-st…</span>
<span class="co">#&gt;  9 &lt;list [0]&gt; 108904      &lt;df[,2] [2… 10               2015-10… The int…</span>
<span class="co">#&gt; 10 &lt;list [0]&gt; 30566007    &lt;df[,2] [2… 10               2015-09… The lun…</span>
<span class="co">#&gt; 11 &lt;list [0]&gt; 3604330583  &lt;df[,2] [3… 128              2015-10… We perf…</span>
<span class="co">#&gt; 12 &lt;list [0]&gt; 8653463     &lt;df[,2] [2… 30               2016-06… Morphin…</span>
<span class="co">#&gt; 13 &lt;list [0]&gt; 6796169358  &lt;df[,2] [3… 257              2016-06… Obesity…</span>
<span class="co">#&gt; 14 &lt;list [0]&gt; 3868574676  &lt;df[,2] [3… 882              2016-02… Accumul…</span>
<span class="co">#&gt; 15 &lt;list [0]&gt; 260800      &lt;df[,2] [3… 55               2016-11… Cancero…</span>
<span class="co">#&gt; # … with 17 more variables: accession &lt;chr&gt;, published &lt;chr&gt;, title &lt;chr&gt;,</span>
<span class="co">#&gt; #   mean_bases_per_run &lt;dbl&gt;, run_count &lt;chr&gt;, center_name &lt;chr&gt;,</span>
<span class="co">#&gt; #   study_type &lt;chr&gt;, insdc &lt;lgl&gt;, sample_count &lt;chr&gt;, alias &lt;chr&gt;,</span>
<span class="co">#&gt; #   attributes &lt;list&gt;, total_bases &lt;chr&gt;, lastupdate &lt;chr&gt;, BioProject &lt;chr&gt;,</span>
<span class="co">#&gt; #   status &lt;chr&gt;, description &lt;chr&gt;, broker_name &lt;chr&gt;</span></pre></body></html></div>
</div>
<div id="what-we-can-do-with-the-srastudies-endpoint" class="section level2">
<h2 class="hasAnchor">
<a href="#what-we-can-do-with-the-srastudies-endpoint" class="anchor"></a>What we can do with the <code>sra/studies</code> endpoint</h2>
<p>Let’s use the <code>refs</code> entity computed above to see what we can request using the <code>sra/studies</code> endpoint.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">ourendpt</span> <span class="kw">=</span> <span class="st">"/sra/studies"</span>
<span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="kw">qualname</span><span class="kw">=</span><span class="no">ref</span>$<span class="no">paths</span><span class="kw">[[</span><span class="no">ourendpt</span>]]$<span class="no">get</span>$<span class="no">parameters</span>$<span class="no">name</span>,
  <span class="no">ref</span>$<span class="no">paths</span><span class="kw">[[</span><span class="no">ourendpt</span>]]$<span class="no">get</span>$<span class="no">parameters</span>$<span class="no">schema</span>))  <span class="co"># wanted datatable but threw warning</span>
<span class="co">#&gt;         qualname          title    type</span>
<span class="co">#&gt; 1              q              Q  string</span>
<span class="co">#&gt; 2           size           Size integer</span>
<span class="co">#&gt; 3         cursor         Cursor  string</span>
<span class="co">#&gt; 4     facet_size     Facet_Size integer</span>
<span class="co">#&gt; 5 include_fields Include_Fields   array</span>
<span class="co">#&gt; 6 exclude_fields Exclude_Fields   array</span>
<span class="co">#&gt; 7         facets         Facets   array</span>
<span class="co">#&gt;                                                                                                                                                                                                                                                                                                                    description</span>
<span class="co">#&gt; 1                                                                                                                                                                                                                          The query, using [lucene query syntax](https://lucene.apache.org/core/3_6_0/queryparsersyntax.html)</span>
<span class="co">#&gt; 2                                                                                                                                                                                                                                                                                                                         &lt;NA&gt;</span>
<span class="co">#&gt; 3 The cursor is used to scroll through results. For a query with more results than `size`, the result will include `cursor` in the result json. Use that value here and re-issue the query. The next set or results will be returned. When no more results are available, the `cursor` will again be empty in the result json.</span>
<span class="co">#&gt; 4                                                                                                                                                                                                           The maximum number of records returned for each facet. This has no effect unless one or more facets are specified.</span>
<span class="co">#&gt; 5                                                                                                                                                                                                                                                               Fields to include in results. The default is to all fields (*)</span>
<span class="co">#&gt; 6                                                                                                                                                                                                                                                   Fields to exclude from results. The default is to not exclude any fields. </span>
<span class="co">#&gt; 7                                                                                                                          A list of strings identifying fields for faceted search results. Simple term faceting is used here, meaning that fields that are short text and repeated across records will be binned and counted.</span>
<span class="co">#&gt;   example exclusiveMaximum default   type</span>
<span class="co">#&gt; 1  cancer               NA    NULL   &lt;NA&gt;</span>
<span class="co">#&gt; 2      10             1000      10   &lt;NA&gt;</span>
<span class="co">#&gt; 3    NULL               NA    NULL   &lt;NA&gt;</span>
<span class="co">#&gt; 4      10               NA      10   &lt;NA&gt;</span>
<span class="co">#&gt; 5       *               NA    NULL string</span>
<span class="co">#&gt; 6    NULL               NA    NULL string</span>
<span class="co">#&gt; 7    NULL               NA    NULL string</span></pre></body></html></div>
<p>The <code>q</code> parameter is used to specify queries in the Lucene query format. We’ll return to that later.</p>
</div>
</div>
<div id="programmatic-approaches-to-inspecting-resources" class="section level1">
<h1 class="hasAnchor">
<a href="#programmatic-approaches-to-inspecting-resources" class="anchor"></a>Programmatic approaches to inspecting resources</h1>
<div id="a-single-run" class="section level2">
<h2 class="hasAnchor">
<a href="#a-single-run" class="anchor"></a>A single run</h2>
<p>Consider the content of the <a href="https://github.com/waldronlab/curatedMetagenomicDataCuration/tree/master/inst/curated">inst/curated folder</a> of the curatedMetagenomicDataCuration repository. This is a collection of tsv files with sample attributes and <code>NCBI_accession</code> numbers for the samples. Each tsv file has a name in format <code>[tag]_[year]</code> where <code>tag</code> is often an author name.</p>
<p>The first folder is called <code>AsnicarF_2017</code>. There are two files: the <code>AsnicarF_2017_metadata.tsv</code> and a python program. Other folders may have only the tsv file.</p>
<p>Let’s get information on a specific sample in this study. We read from the tsv file in github the SRR number of the first row: SRR4052021. This is a token that can be used with the /sra/runs endpoint of OmicIDX.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">pref</span> <span class="kw">=</span> <span class="st">"https://api.omicidx.cancerdatasci.org/"</span>
<span class="no">do_one</span> <span class="kw">=</span> <span class="st">"sra/runs/SRR4052021"</span>
<span class="no">lk1</span> <span class="kw">=</span> <span class="kw pkg">jsonlite</span><span class="kw ns">::</span><span class="fu"><a href="https://jeroen.cran.dev/jsonlite/reference/fromJSON.html">fromJSON</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">pref</span>, <span class="no">do_one</span>))
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">lk1</span>)
<span class="co">#&gt;  [1] "cluster_name"          "static_data_available" "study"                </span>
<span class="co">#&gt;  [4] "total_spots"           "avg_length"            "sample_accession"     </span>
<span class="co">#&gt;  [7] "experiment_accession"  "received"              "accession"            </span>
<span class="co">#&gt; [10] "published"             "sample"                "study_accession"      </span>
<span class="co">#&gt; [13] "size"                  "experiment"            "insdc"                </span>
<span class="co">#&gt; [16] "load_done"             "is_public"             "alias"                </span>
<span class="co">#&gt; [19] "qualities"             "attributes"            "total_bases"          </span>
<span class="co">#&gt; [22] "lastupdate"            "status"</span></pre></body></html></div>
<p>So there is a straightforward way to get detailed metadata for an SRR. Massaging the output here, to get that which is useful for metagenomic analysis, is an exercise.</p>
</div>
<div id="a-study" class="section level2">
<h2 class="hasAnchor">
<a href="#a-study" class="anchor"></a>A study</h2>
<p>We encapsulated code we used above to get metadata on the <code>AsnicarF_2017</code> study in the function <code>samp_atts_by_study</code>, that we’ll show and demonstrate here. We determined the SRP number from the call above.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>({
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">md4mg</span>)
})
<span class="co">#&gt; Warning: package 'jsonlite' was built under R version 4.0.2</span>
<span class="no">samp_atts_by_study</span>
<span class="co">#&gt; function (srp = "SRP082656", size = 500, cursor = NULL, ...) </span>
<span class="co">#&gt; {</span>
<span class="co">#&gt;     base_url = "https://api.omicidx.cancerdatasci.org/sra/studies/%s/runs?size=%s"</span>
<span class="co">#&gt;     if (!is.null(cursor)) </span>
<span class="co">#&gt;         base_url = paste0(base_url, "&amp;cursor=", cursor)</span>
<span class="co">#&gt;     qstr = sprintf(base_url, srp, size)</span>
<span class="co">#&gt;     md = try(jsonlite::fromJSON(txt = qstr, flatten = TRUE))</span>
<span class="co">#&gt;     if (size == 0) </span>
<span class="co">#&gt;         return(md)</span>
<span class="co">#&gt;     if (inherits(md, "try-error")) </span>
<span class="co">#&gt;         stop("could not parse JSON")</span>
<span class="co">#&gt;     metadata = as_tibble(md$hits)</span>
<span class="co">#&gt;     if (inherits(metadata, "try-error")) </span>
<span class="co">#&gt;         stop("check srp")</span>
<span class="co">#&gt;     sampatts = metadata %&gt;% dplyr::select(accession, sample.attributes) %&gt;% </span>
<span class="co">#&gt;         tidyr::unnest(sample.attributes) %&gt;% dplyr::filter(tag != </span>
<span class="co">#&gt;         "BioSampleModel") %&gt;% tidyr::pivot_wider(names_from = "tag", </span>
<span class="co">#&gt;         values_from = "value")</span>
<span class="co">#&gt;     attr(sampatts, "cursor") = md$cursor</span>
<span class="co">#&gt;     sampatts</span>
<span class="co">#&gt; }</span>
<span class="co">#&gt; &lt;bytecode: 0x562d239522d0&gt;</span>
<span class="co">#&gt; &lt;environment: namespace:md4mg&gt;</span>
<span class="no">sa_asnicar</span> <span class="kw">=</span> <span class="fu"><a href="https://vjcitn.github.io/md4mg/reference/samp_atts_by_study.html">samp_atts_by_study</a></span>()
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">sa_asnicar</span>)
<span class="co">#&gt; [1] 28 11</span>
<span class="kw pkg">DT</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html">datatable</a></span>(<span class="no">sa_asnicar</span>)</pre></body></html></div>
<div id="htmlwidget-e34ee1512293350ee042" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-e34ee1512293350ee042">{"x":{"filter":"none","data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28"],["SRR4052021","SRR4052022","SRR4052023","SRR4052024","SRR4052025","SRR4052026","SRR4052027","SRR4052028","SRR4052029","SRR4052030","SRR4052031","SRR4052032","SRR4052033","SRR4052034","SRR4052035","SRR4052036","SRR4052037","SRR4052038","SRR4052039","SRR4052040","SRR4052041","SRR4052042","SRR4052043","SRR4052044","SRR4100706","SRR4100707","SRR4100708","SRR4100709"],["2014","2014","2014","2014","2015","2014","2014","2015","2014","2014","2014","2014","2014","2015","2014","2014","2015","2014","2015","2014","2014","2015","2014","2014","2015","2015","2015","2015"],["Human stool","Human stool","Human stool","Human stool","Human stool","Human stool","Human stool","Human stool","Human milk","Human milk","Human milk","Human milk","Human stool","Human milk","Human milk","Human milk","Human milk","Human stool","Human stool","Human stool","Human stool","Human stool","Human stool","Human stool","Human stool","Human stool","Human stool","Human stool"],["Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat","Human-associated habitat"],["Human infant stool","Human infant stool","Human mother stool","Human mother stool","Human mother stool","Human mother stool","Human mother stool","Human mother stool","Human mother milk","Human mother milk","Human mother milk","Human mother milk","Human infant stool","Human mother milk","Human mother milk","Human mother milk","Human mother milk","Human infant stool","Human infant stool","Human infant stool","Human infant stool","Human infant stool","Human mother stool","Human mother stool","Human mother stool","Human mother stool","Human mother stool","Human mother stool"],["Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento","Italy: Trento"],["Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens","Homo sapiens"],["46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E","46.0667 N 11.1167 E"],["T1","T1","T1","T1","T2","T1","T2","T3","T1","T1","T1","T1","T1","T2","T1","T2","T3","T1","T2","T1","T2","T3","T1","T1","T2","T3","T2","T3"],["female","male","female","female","female","female","female","female","female","female","female","female","male","female","female","female","female","male","male","male","male","male","female","female","male","male","female","female"],["1","2","3","4","4","5","5","5","1","2","3","4","3","4","5","5","5","4","4","5","5","5","1","2","4","5","4","5"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>accession<\/th>\n      <th>collection_date<\/th>\n      <th>env_biome<\/th>\n      <th>env_feature<\/th>\n      <th>env_material<\/th>\n      <th>geo_loc_name<\/th>\n      <th>host<\/th>\n      <th>lat_lon<\/th>\n      <th>timepoint<\/th>\n      <th>gender<\/th>\n      <th>couple<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"order":[],"autoWidth":false,"orderClasses":false,"columnDefs":[{"orderable":false,"targets":0}]}},"evals":[],"jsHooks":[]}</script><p>One could say that this is the uncurated sample metadata for this study. Of some concern is the fact that there are 28 records here but only 25 records in the tsv file.</p>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:stvjc@channing.harvard.edu" class="email">stvjc@channing.harvard.edu</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p><a href="mailto:seandavi@gmail.com" class="email">seandavi@gmail.com</a><a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Vincent Carey, Sean Davis.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
